### DISCUSSION TITLE
Could you share any papers/books that provide underlying explanations for this project?

### DISCUSSION THREAD

**Discussion by nicks64** [ggml-org/llama.cpp#6879] [General]
Hi i am looking to contribute to this project. I am just getting started... and I am looking for papers that explain the logic/concept of this open source project. Could you share any and everything that i should look into before I jump in? 

I know that attention is all you need is a good start.

Thank you in advance

**Comment by wilderfield** [comment]
Reading papers may help you understand the mathematical theory, but likely won't help you understand all the infrastructure here and the focus on performance optimization.

I recommend building main.exe in debug mode, and stepping through the callflow with gdb.

There was also a discussion where someone shared their learnings.

**Reply by nicks64** [reply]
do you have any books in mind or any research papers... because there is a basis

**Reply by marcingomulkiewicz** [reply]
"Attention is all you need" ? :wink:

**Reply by nicks64** [reply]
any more? but thank you

**Comment by ngxson** [comment]
If you are looking to contribute, the best way to start is by actually using it as a ["daily driver"](https://www.quora.com/Why-do-smartphone-YouTubers-refer-to-their-phones-as-a-daily-driver) tool.

While "Attention is all you need" is a good start, you will find that it's very hard to actually apply the idea directly into the code. There's many other part in the project that you need to understand, for example backend (CPU/GPU) implementation, kv cache, gguf format,...

For that reason, it's better to just use it, maybe also read all the `docs`, have a look on some merged PR to have a sense of what kind of contribution you can make.

And remember, only make contributions when it's good for community - not just good for you. This video pretty much sum that up: https://www.youtube.com/watch?v=5nY_cy8zcO4

**Reply by phymbert** [reply]
Regarding LLM, this video series helped me a lot:

https://youtu.be/eMlx5fFNoYc

**Reply by nicks64** [reply]
what do you mean as a daily driver tool?

**Reply by ngxson** [reply]
I mean using it as a normal everyday use case. A good example can be #6000 where many users (including me) use llama.cpp everyday, find that there's a missing feature, and then we make contributions to implement it.

What I'm trying to say here is that good contributions are the one solve real-life problems. I'm just pointing that out because there're currently many not-so-good PRs just because someone wants to "contribute", rather than "actually resolve a problem".
